pkg/apis/core/v1/defaults_test.go

```diff
@@ -379,6 +379,571 @@ func testPodDefaults(t *testing.T, featuresEnabled bool) {
+				},
+				Limits: v1.ResourceList{
+					"cpu":    resource.MustParse("2m"),
+					"memory": resource.MustParse("1Mi"),
+				},
+			},
+		},
+	}, {
+		name:                     "pod requests=unset limits=set, container requests=unset limits=set",
+		podLevelResourcesEnabled: true,
+		containers: []v1.Container{
+			{
+				Resources: v1.ResourceRequirements{
+					Limits: v1.ResourceList{
+						"cpu":    resource.MustParse("2m"),
+						"memory": resource.MustParse("1Mi"),
+					},
+				},
+			}, {
+				Resources: v1.ResourceRequirements{
+					Limits: v1.ResourceList{
+						"cpu":    resource.MustParse("1m"),
+						"memory": resource.MustParse("5Mi"),
+					},
+				},
+			},
+		},
+		podResources: &v1.ResourceRequirements{
+			Limits: v1.ResourceList{
+				"cpu":    resource.MustParse("5m"),
+				"memory": resource.MustParse("7Mi"),
+			},
+		},
+		expectedPodSpec: v1.PodSpec{
+			Resources: &v1.ResourceRequirements{
+				Requests: v1.ResourceList{
+					"cpu":    resource.MustParse("3m"),
+					"memory": resource.MustParse("6Mi"),
+				},
+				Limits: v1.ResourceList{
+					"cpu":    resource.MustParse("5m"),
+					"memory": resource.MustParse("7Mi"),
+				},
+			},
+			Containers: []v1.Container{
+				{
+					Resources: v1.ResourceRequirements{
+						Requests: v1.ResourceList{
+							"cpu":    resource.MustParse("2m"),
+							"memory": resource.MustParse("1Mi"),
+						},
+						Limits: v1.ResourceList{
+							"cpu":    resource.MustParse("2m"),
+							"memory": resource.MustParse("1Mi"),
+						},
+					},
+				}, {
+					Resources: v1.ResourceRequirements{
+						Requests: v1.ResourceList{
+							"cpu":    resource.MustParse("1m"),
+							"memory": resource.MustParse("5Mi"),
+						},
+						Limits: v1.ResourceList{
+							"cpu":    resource.MustParse("1m"),
+							"memory": resource.MustParse("5Mi"),
+						},
+					},
+				},
+			},
+		},
+	}, {
+		name:                     "pod requests=unset limits=set, container requests=set limits=unset",
+		podLevelResourcesEnabled: true,
+		podResources: &v1.ResourceRequirements{
+			Limits: v1.ResourceList{
+				"cpu":    resource.MustParse("4m"),
+				"memory": resource.MustParse("8Mi"),
+			},
+		},
+		containers: []v1.Container{
+			{
+				Resources: v1.ResourceRequirements{
+					Requests: v1.ResourceList{
+						"cpu":    resource.MustParse("1m"),
+						"memory": resource.MustParse("1Mi"),
+					},
+				},
+			}, {
+				Resources: v1.ResourceRequirements{
+					Requests: v1.ResourceList{
+						"cpu":    resource.MustParse("1m"),
+						"memory": resource.MustParse("2Mi"),
+					},
+				},
+			},
+		},
+		expectedPodSpec: v1.PodSpec{
+			Resources: &v1.ResourceRequirements{
+				Requests: v1.ResourceList{
+					"cpu":    resource.MustParse("2m"),
+					"memory": resource.MustParse("3Mi"),
+				},
+				Limits: v1.ResourceList{
+					"cpu":    resource.MustParse("4m"),
+					"memory": resource.MustParse("8Mi"),
+				},
+			},
+			Containers: []v1.Container{
```

Comment:
from what I can tell this is the only test that tests a non homogeneously set group of containers (where they have different permutations of requests/limits set). I am curious about tests with two containers A and B:
A limits !request, B requests !limit
A limits request, B requests !limit
A limits request, B !requests limit // this is covered here
with and without all pod resources (limits, requests, limits & requests)

Reply:
Tests for these cases are explicitly not needed because unit tests of PodRequests covers all these cases. Whether 1 container or multiple containers have requests/limits set/unset, the defaulting logic just uses the calculation PodRequests returns. I will still add a few more test cases in the follow up PR.